test/quantization/test_quantize_script.py

```diff
@@ -2348,10 +2339,81 @@ def forward(self, x):
                 return self.fc(x), size1, size2
 
         model = torch.jit.script(M()).eval()
-        data = torch.rand((1, 5), dtype=torch.float)
         qconfig_dict = {'': default_dynamic_qconfig}
 
-        model = quantize_dynamic_script(model, qconfig_dict, [data])
+        model = quantize_dynamic_script(model, qconfig_dict)
         FileCheck().check("quantized::linear_dynamic") \
                    .check_not("aten::_choose_qparams_per_tensor") \
                    .run(model.graph)
+
+    @override_qengines
+    def test_dynamic_shared_weights(self):
+        class myMod(torch.nn.Module):
+            def __init__(self, weight):
+                super().__init__()
+                self.linear = nn.Linear(5, 5)
+                self.linear.weight = weight
+
+            def forward(self, x):
+                return self.linear(x)
+
+        class DynamicModel(torch.nn.Module):
+            def __init__(self):
+                super(DynamicModel, self).__init__()
+                self.weight = torch.nn.Parameter(torch.ones(5, 5))
+                self.mod1 = myMod(self.weight)
+
+            def forward(self, x):
+                y = self.mod1(x)
+                z = torch.nn.functional.linear(y, self.weight)
+                return z
+
+        model = torch.jit.script(DynamicModel()).eval()
+        data = torch.randn(5, 5, dtype=torch.float)
+        for op in ['mod1', '']:
+            qconfig_dict = {op: default_dynamic_qconfig}
+            model_graph = quantize_dynamic_script(model, qconfig_dict)
+            out_graph = model_graph(data)
+
+            # Explicitly call forward on model before convert
+            m = prepare_dynamic_script(model, qconfig_dict)
+            m(data)
+            m = convert_dynamic_script(m, debug=False)
+            out_ref = model_graph(data)
+            self.assertEqual(out_graph, out_ref)
+
+    @override_qengines
+    def test_dynamic_with_if(self):
+        class Res(torch.nn.Module):
+            def __init__(self):
+                super(Res, self).__init__()
+                self.weight = torch.nn.Parameter(torch.ones(5, 5))
+
+            def forward(self, x, cond):
+                # type: (Tensor, bool) -> Tensor
+                if cond:
+                    return torch.nn.functional.linear(x, self.weight)
+                else:
+                    return torch.nn.functional.linear(x, self.weight)
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                super(M, self).__init__()
+                self.res1 = Res()
+                self.res2 = Res()
+
+            def forward(self, x):
+                x = self.res1(x, True)
```

Comment:
For this test to be more useful, please test with dynamic inputs to forward. How about 
``` python 
def forward(self, x):
  cond=  torch.norm(x) > 10
  x = self.res1(x, cond)

Reply:
There seems to be an issue with this in general (not related to my PR). Will debug later.